### PR TITLE
App platform documentation updates

### PR DESCRIPTION
**Pull request by andrewsomething**
This documents the image source (#565), jobs component (#566),  domain changes (#572), and service internal ports (#570).

~Note: #570 has not yet been merged.~

### CODE REVIEW

File: docs/resources/app.md [hunk 1]
```diff
@@ -121,7 +123,14 @@ The following arguments are supported:
...
  - `name` - (Required) The name of the app. Must be unique across all apps in the same account.
  - `region` - The slug for the DigitalOcean data center region hosting the app.
- - `domains` - A list of hostnames where the application will be available.
+ - `domain` - Describes a domain where the application will be made available
+     * `name` - The hostname for the domain
+     * `type` - The domain type, which can be one of the following:
+         - `DEFAULT`: The default .ondigitalocean.app domain assigned to this app
+         - `PRIMARY`: The primary domain for this app that is displayed as the default in the control panel, used in bindable environment variables, and any other places that reference an app's live URL. Only one domain may be set as primary.
+         - `ALIAS`: A non-primary domain
+     * `wildcard` - A boolean indicating whether the domain includes all sub-domains, in addition to the given domain
```
Comment (ChiefMateStarbuck) [inline docs/resources/app.md:132]:
```suggestion
     * `wildcard` - A boolean indicating whether the domain includes all sub-domains, in addition to the given domain.
```

File: docs/resources/app.md [hunk 2]
```diff
@@ -141,17 +150,23 @@ A `service` can contain:
...
   - `repo` - The name of the repo in the format `owner/repo`.
   - `branch` - The name of the branch to use.
   - `deploy_on_push` - Whether to automatically deploy new commits made to the repo.
-* `gitlab` - A Gitlab repo to use as the component's source. DigitalOcean App Platform must have [access to the repository](https://cloud.digitalocean.com/apps/gitlab/install). Only one of `git`, `github` or `gitlab`  may be set.
+* `gitlab` - A Gitlab repo to use as the component's source. DigitalOcean App Platform must have [access to the repository](https://cloud.digitalocean.com/apps/gitlab/install). Only one of `git`, `github`, `gitlab`, or `image` may be set.
   - `repo` - The name of the repo in the format `owner/repo`.
   - `branch` - The name of the branch to use.
   - `deploy_on_push` - Whether to automatically deploy new commits made to the repo.
+* `image` - An image to use as the component's source. Only one of `git`, `github`, `gitlab`, or `image` may be set.
+  - `registry_type` - The registry type. One of `DOCR` (DigitalOcean container registry) or `DOCKER_HUB`
```
Comment (ChiefMateStarbuck) [inline docs/resources/app.md:166]:
```suggestion
  - `registry_type` - The registry type. One of `DOCR` (DigitalOcean container registry) or `DOCKER_HUB`.
```

File: docs/resources/app.md [hunk 3]
```diff
@@ -167,6 +182,36 @@ A `service` can contain:
   - `success_threshold` - The number of successful health checks before considered healthy.
   - `failure_threshold` - The number of failed health checks before considered unhealthy.
 
+A `static_site` can contain:
+
+* `name` - The name of the component
```
Comment (ChiefMateStarbuck) [inline docs/resources/app.md:187]:
```suggestion
* `name` - The name of the component.
```

File: docs/resources/app.md [hunk 4]
```diff
@@ -167,6 +182,36 @@ A `service` can contain:
...
+* `build_command` - An optional build command to run while building this component from source.
+* `dockerfile_path` - The path to a Dockerfile relative to the root of the repo. If set, overrides usage of buildpacks.
+* `source_dir` - An optional path to the working directory to use for the build.
+* `environment_slug` - An environment slug describing the type of this app.
+* `output_dir` - An optional path to where the built assets will be located, relative to the build context. If not set, App Platform will automatically scan for these directory names: `_static`, `dist`, `public`.
+* `index_document` - The name of the index document to use when serving this static site.
+* `error_document` - The name of the error document to use when serving this static site.
+* `catchall_document` - The name of the document to use as the fallback for any requests to documents that are not found when serving this static site.
+* `git` - A Git repo to use as the component's source. The repository must be able to be cloned without authentication.  Only one of `git`, `github` or `gitlab`  may be set
```
Comment (ChiefMateStarbuck) [inline docs/resources/app.md:196]:
```suggestion
* `git` - A Git repo to use as the component's source. The repository must be able to be cloned without authentication.  Only one of `git`, `github` or `gitlab`  may be set.
```

File: docs/resources/app.md [hunk 5]
```diff
@@ -180,49 +225,61 @@ A `worker` can contain:
...
   - `repo` - The name of the repo in the format `owner/repo`.
   - `branch` - The name of the branch to use.
   - `deploy_on_push` - Whether to automatically deploy new commits made to the repo.
-* `gitlab` - A Gitlab repo to use as the component's source. DigitalOcean App Platform must have [access to the repository](https://cloud.digitalocean.com/apps/gitlab/install). Only one of `git`, `github` or `gitlab`  may be set.
+* `gitlab` - A Gitlab repo to use as the component's source. DigitalOcean App Platform must have [access to the repository](https://cloud.digitalocean.com/apps/gitlab/install). Only one of `git`, `github`, `gitlab`, or `image` may be set.
   - `repo` - The name of the repo in the format `owner/repo`.
   - `branch` - The name of the branch to use.
   - `deploy_on_push` - Whether to automatically deploy new commits made to the repo.
+* `image` - An image to use as the component's source. Only one of `git`, `github`, `gitlab`, or `image` may be set.
+  - `registry_type` - The registry type. One of `DOCR` (DigitalOcean container registry) or `DOCKER_HUB`
```
Comment (ChiefMateStarbuck) [inline docs/resources/app.md:237]:
```suggestion
  - `registry_type` - The registry type. One of `DOCR` (DigitalOcean container registry) or `DOCKER_HUB`.
```

File: docs/resources/app.md [hunk 6]
```diff
@@ -180,49 +225,61 @@ A `worker` can contain:
...
```
Comment (ChiefMateStarbuck) [inline docs/resources/app.md:274]:
```suggestion
  - `registry_type` - The registry type. One of `DOCR` (DigitalOcean container registry) or `DOCKER_HUB`.
```

File: docs/data-sources/app.md [hunk 7]
```diff
@@ -48,17 +48,23 @@ A `service` can contain:
...
   - `repo` - The name of the repo in the format `owner/repo`.
   - `branch` - The name of the branch to use.
   - `deploy_on_push` - Whether to automatically deploy new commits made to the repo.
-* `gitlab` - A Gitlab repo to use as component's source. Only one of `git`, `github` or `gitlab` may be set.
+* `gitlab` - A Gitlab repo to use as component's source. Only one of `git`, `github`, `gitlab`, or `image` may be set.
   - `repo` - The name of the repo in the format `owner/repo`.
   - `branch` - The name of the branch to use.
   - `deploy_on_push` - Whether to automatically deploy new commits made to the repo.
+* `image` - An image to use as the component's source. Only one of `git`, `github`, `gitlab`, or `image` may be set.
+  - `registry_type` - The registry type. One of `DOCR` (DigitalOcean container registry) or `DOCKER_HUB`
```
Comment (ChiefMateStarbuck) [inline docs/data-sources/app.md:64]:
```suggestion
  - `registry_type` - The registry type. One of `DOCR` (DigitalOcean container registry) or `DOCKER_HUB`.
```

File: docs/data-sources/app.md [hunk 8]
```diff
@@ -74,6 +80,36 @@ A `service` can contain:
   - `success_threshold` - The number of successful health checks before considered healthy.
   - `failure_threshold` - The number of failed health checks before considered unhealthy.
 
+A `static_site` can contain:
+
+* `name` - The name of the component
```
Comment (ChiefMateStarbuck) [inline docs/data-sources/app.md:85]:
```suggestion
* `name` - The name of the component.
```

File: docs/data-sources/app.md [hunk 9]
```diff
@@ -84,52 +120,64 @@ A `worker` can contain:
...
   - `repo` - The name of the repo in the format `owner/repo`.
   - `branch` - The name of the branch to use.
   - `deploy_on_push` - Whether to automatically deploy new commits made to the repo.
-* `gitlab` - A Gitlab repo to use as component's source. Only one of `git`, `github` or `gitlab` may be set.
+* `gitlab` - A Gitlab repo to use as component's source. Only one of `git`, `github`, `gitlab`, or `image` may be set.
   - `repo` - The name of the repo in the format `owner/repo`.
   - `branch` - The name of the branch to use.
   - `deploy_on_push` - Whether to automatically deploy new commits made to the repo.
+* `image` - An image to use as the component's source. Only one of `git`, `github`, `gitlab`, or `image` may be set.
+  - `registry_type` - The registry type. One of `DOCR` (DigitalOcean container registry) or `DOCKER_HUB`
```
Comment (ChiefMateStarbuck) [inline docs/data-sources/app.md:135]:
```suggestion
  - `registry_type` - The registry type. One of `DOCR` (DigitalOcean container registry) or `DOCKER_HUB`.
```

File: docs/data-sources/app.md [hunk 10]
```diff
@@ -84,52 +120,64 @@ A `worker` can contain:
...
+* `run_command` - An optional run command to override the component's default.
 * `environment_slug` - An environment slug describing the type of this app.
-* `output_dir` - An optional path to where the built assets will be located, relative to the build context. If not set, App Platform will automatically scan for these directory names: `_static`, `dist`, `public`.
-* `index_document` - The name of the index document to use when serving this static site.
-* `error_document` - The name of the error document to use when serving this static site.
-* `catchall_document` - The name of the document to use as the fallback for any requests to documents that are not found when serving this static site.
-* `git` - A Git repo to use as component's source. Only one of `git`, `github` or `gitlab` may be set.
+* `instance_size_slug` - The instance size to use for this component.
+* `instance_count` - The amount of instances that this component should be scaled to.
+* `git` - A Git repo to use as the component's source. The repository must be able to be cloned without authentication.  Only one of `git`, `github` or `gitlab`  may be set
```
Comment (ChiefMateStarbuck) [inline docs/data-sources/app.md:160]:
```suggestion
* `git` - A Git repo to use as the component's source. The repository must be able to be cloned without authentication.  Only one of `git`, `github` or `gitlab`  may be set.
```

File: docs/data-sources/app.md [hunk 11]
```diff
@@ -84,52 +120,64 @@ A `worker` can contain:
...
   - `repo` - The name of the repo in the format `owner/repo`.
   - `branch` - The name of the branch to use.
   - `deploy_on_push` - Whether to automatically deploy new commits made to the repo.
-* `gitlab` - A Gitlab repo to use as component's source. Only one of `git`, `github` or `gitlab` may be set.
+* `gitlab` - A Gitlab repo to use as the component's source. DigitalOcean App Platform must have [access to the repository](https://cloud.digitalocean.com/apps/gitlab/install). Only one of `git`, `github`, `gitlab`, or `image` may be set.
   - `repo` - The name of the repo in the format `owner/repo`.
   - `branch` - The name of the branch to use.
   - `deploy_on_push` - Whether to automatically deploy new commits made to the repo.
+* `image` - An image to use as the component's source. Only one of `git`, `github`, `gitlab`, or `image` may be set.
+  - `registry_type` - The registry type. One of `DOCR` (DigitalOcean container registry) or `DOCKER_HUB`
```
Comment (ChiefMateStarbuck) [inline docs/data-sources/app.md:172]:
```suggestion
  - `registry_type` - The registry type. One of `DOCR` (DigitalOcean container registry) or `DOCKER_HUB`.
```

**Review by ChiefMateStarbuck — CHANGES_REQUESTED** [review]
Overall, looks great 🤠

**Review by bentranter — DISMISSED** [review]
LGTM!

File: docs/resources/app.md [hunk 12]
```diff
@@ -121,7 +123,14 @@ The following arguments are supported:
 * `spec` - (Required) A DigitalOcean App spec describing the app.
  - `name` - (Required) The name of the app. Must be unique across all apps in the same account.
  - `region` - The slug for the DigitalOcean data center region hosting the app.
- - `domains` - A list of hostnames where the application will be available.
+ - `domain` - Describes a domain where the application will be made available
+     * `name` - The hostname for the domain
+     * `type` - The domain type, which can be one of the following:
+         - `DEFAULT`: The default .ondigitalocean.app domain assigned to this app
```
[the remark this answers is not in the document]
Comment (andrewsomething) [reply docs/resources/app.md]: When you create an app, you get an automated sub-domain at `$APP.ondigitalocean.app`. Just reusing the product docs wording here:

https://www.digitalocean.com/docs/app-platform/references/app-specification-reference/#ref-domains-type

**Review by ChiefMateStarbuck — APPROVED** [review]
LGTM

**Pull request merged**